### PR TITLE
[Dem-Structural] Adding 2.5D Dem-Fem coupling features

### PR DESCRIPTION
**Pull request by ipouplana**
Summary of changes in this PR ordered by application:

**StructuralMechanics Application** (please @KratosMultiphysics/structural-mechanics take a look at it)
- New 2.5D solid element (essentially a 2D element that works with a 3D CL)
- New test of the 2.5D solid element (0.1s in FullDebug with 1 thread)
- New process to impose Z strain values into the member variable of the 2.5D solid element (just created for the test). Note that for the moment this is the simplest possible process to imposes constant Z strain values, but we could generalize it in the future to impose Z strain functions if it may be of any interest.
- New double variable IMPOSED_Z_STRAIN_VALUE
- New bool variable IMPOSED_Z_STRAIN_OPTION

**DEM Application**
- New 2.5D LineLoadDEMFEM condition
- New Control Module 2.5D utility
- RigidEdge3D adapted for the 2.5D DEM-FEM coupling
- GluedToWallScheme adapted for the 2.5D DEM-FEM coupling
- Control Module Process minor revisions
- Warning removal

### CODE REVIEW

**Review by philbucher — CHANGES_REQUESTED** [review]
Hi @ipouplana 
I have some comments to make but I am traveling today, I will review tomorrow

File: kratos/sources/variables.cpp
Comment (philbucher) [inline kratos/sources/variables.cpp]: I would not move those to the core, to me are not general enough

File: applications/StructuralMechanicsApplication/tests/test_solid_2p5d_element.py
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/test_solid_2p5d_element.py]: I would prefer since this is a standard test (i.e. has ProjectParameters etc) to add it in the standard way (i.e. in the test-factory)

File: applications/StructuralMechanicsApplication/python_scripts/impose_z_strain_process.py
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/impose_z_strain_process.py]: I recommend simplifying this like I did [here](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/StructuralMechanicsApplication/python_scripts/postprocess_eigenvalues_process.py#L13-L39)

this way you are sure all the functions of your c++ process are being called

File: applications/StructuralMechanicsApplication/python_scripts/impose_z_strain_process.py
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/impose_z_strain_process.py]: why don't you directly pass the settings from outside?

File: applications/StructuralMechanicsApplication/custom_processes/impose_z_strain_process.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_processes/impose_z_strain_process.hpp]:
```suggestion
    ModelPart& mrModelPart;

```

File: applications/StructuralMechanicsApplication/custom_processes/impose_z_strain_process.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_processes/impose_z_strain_process.hpp]: Please add some documentation and separate into header & source

File: applications/StructuralMechanicsApplication/custom_processes/impose_z_strain_process.hpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_processes/impose_z_strain_process.hpp]: we require c++11 hence I would delete this comment

File: applications/StructuralMechanicsApplication/custom_elements/z_strain_driven_2p5_small_displacement.h
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/z_strain_driven_2p5_small_displacement.h]:
```suggestion
    void Initialize(const ProcessInfo rCurrentProcessInfo) override;
```

File: applications/StructuralMechanicsApplication/custom_elements/z_strain_driven_2p5_small_displacement.h
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/z_strain_driven_2p5_small_displacement.h]:
```suggestion
    ~ZStrainDriven2p5DSmallDisplacement() override = default;
```

File: applications/StructuralMechanicsApplication/custom_elements/z_strain_driven_2p5_small_displacement.h
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/z_strain_driven_2p5_small_displacement.h]: please remove commented code

File: applications/StructuralMechanicsApplication/custom_elements/z_strain_driven_2p5_small_displacement.h
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/z_strain_driven_2p5_small_displacement.h]: some of these are not used / only used in the cpp, please check

File: applications/StructuralMechanicsApplication/custom_elements/z_strain_driven_2p5_small_displacement.cpp
Comment (philbucher) [inline applications/StructuralMechanicsApplication/custom_elements/z_strain_driven_2p5_small_displacement.cpp]: I would remove it, see comment in the header file

**Review by philbucher — COMMENTED** [review]
I gave my comments from the side of StructuralMechanics

@KratosMultiphysics/structural-mechanics I have not reviewed the new element in detail

**Comment by ipouplana** [comment]
Hi @philbucher, thanks for the review. I just found some problems in a 3D DEM-FEM test case, and so I want to first make sure that everything works. After that, I will try to address all your comments.

File: kratos/sources/variables.cpp
Comment (ipouplana) [reply kratos/sources/variables.cpp]: The thing is that we need those variables in a problem coupling the Structural Mechanics Application with the DEM Application, and there were no similar variables in Structural Mechanics that we could use. Thereby, in order to avoid repeating the same variable in 2 applications, we decided to move them to the core. 
In any case, if you have a better alternative, please let us know.

Comment (philbucher) [reply kratos/sources/variables.cpp]: I would actually repeat the variable, to me this would not be a problem. Since #1468 it is possible to define the same variable in multiple applications. 

As I said I think this is a quite specific variable that is not general enough for the Core
The Core already has many variables that were only put there because it was previously not possible to define the same variable in several applications

Since this is my opinion/suggestion and since it is not crucial, I will not block because of this.
I just would like to not bloat the Core even more

Comment (ipouplana) [reply kratos/sources/variables.cpp]: Ok, then I'll try defining those variables in both applications ;)

File: applications/StructuralMechanicsApplication/tests/solid_2p5d_test/solid_2p5d_parameters.json
Comment (philbucher) [inline applications/StructuralMechanicsApplication/tests/solid_2p5d_test/solid_2p5d_parameters.json]:
```suggestion
            "solver_type"         : "EigenSolversApplication.sparse_lu"
```

this one is faster, plus the ExternalSolvers will be replaced by the EigenSolvers in the long run

**Review by philbucher — COMMENTED** [review]
From the point of StructuralMechanics I am ok, looks good :+1: (one comment abt the linera-solver)
I will approve once the other app-maintainers also approved

File: applications/StructuralMechanicsApplication/tests/solid_2p5d_test/solid_2p5d_parameters.json
Comment (ipouplana) [reply applications/StructuralMechanicsApplication/tests/solid_2p5d_test/solid_2p5d_parameters.json]: But then we need to compile now the EigenSolversApplication ??

Comment (philbucher) [reply applications/StructuralMechanicsApplication/tests/solid_2p5d_test/solid_2p5d_parameters.json]: I would recommend so, it has more capabilities than the EigenSolversApp
We are slowly switching

Since ~1week the tests in Structural by default use the EigenSolversApp

Comment (philbucher) [reply applications/StructuralMechanicsApplication/tests/solid_2p5d_test/solid_2p5d_parameters.json]: actually you could remove the specification of the linear solver completely and use the default
Then it would also work locally if you don't have the EigenSolversApp compiled

Comment (ipouplana) [reply applications/StructuralMechanicsApplication/tests/solid_2p5d_test/solid_2p5d_parameters.json]: Ok, I'll do that. The thing is that then it skips the test...

Comment (philbucher) [reply applications/StructuralMechanicsApplication/tests/solid_2p5d_test/solid_2p5d_parameters.json]: Ok thats true, if you run it locally without having the EigenSolversApp it will skip it

**Comment by ipouplana** [comment]
Ok, now I am trying to fix compilation errors due to PR #6339 ... Please, do not merge it yet 

(see issue [6525](https://github.com/KratosMultiphysics/Kratos/issues/6525))

**Comment by ipouplana** [comment]
It now compiles ok!

**Review by salvalatorre — APPROVED** [review]
Good job!

**Comment by ipouplana** [comment]
@philbucher could you approve it now ?

**Comment by salvalatorre** [comment]
He couldn't in the end...

**Review by maceligueta — APPROVED** [review]

**Comment by ipouplana** [comment]
Ping @philbucher

**Review by philbucher — APPROVED** [review]
Sorry for the delayed response, I am on vacation

**Pull request merged**
